Manual/ModuleSystem.lean

```diff
@@ -0,0 +1,132 @@
+For proof fields, `by` already does the trick.
+```
+module
+
+def myInternalHelper (x : Nat) : String := ...
+
+public instance : ToString Nat where
+  toString x := private myInternalHelper x
+```
+
+## Import Visibility
+
+The basic form `public import M` makes the public scope of `M` available in the public scope of the current module and discards the private scope.
+Without `public`, the public scope of `M` is instead imported into the private scope.
+The import thus is irrelevant to downstream modules and ignored by them.
+
+`import all M` behaves like `import M` but additional imports the private scope of `M` into the private scope of the current module.
+This is only allowed if `M` and the current module live in the same library as its main purpose is to allow for separating definitions and proofs into separate modules for internal organization of a library.
+```
+-- Module M.Defs
+module
+
+public def f : Nat := 0
+```
+```
+-- Module M.Lemmas
+module
+
+import all M.Defs
+
+public theorem f_eq_zero : f = 0 :=
+  -- may access body of `f` imported into the private scope
+  rfl
+```
+Note that the imported private scope includes private `import`s of `M`, including nested `import all`s that then are interpreted likewise.
+That is, the set of private scopes accessible to the current module is the transitive closure of `import all` declarations.
+Note also that `import all` is more powerful than the basic `import` before the module system as it allows for accessing imported private definitions directly by name, as if they were defined in the current module.
+Thus another use case of `import all` is to make declarations available that need to be used in multiple modules but should not leak outside the current library.
+
+`public import all M` behaves like `public import M` followed by `import all M`, i.e. the `all` modifier becomes irrelevant for downstream modules.
+
+# The `meta` Phase
+
+When it comes to actual code execution, there is no point to a definition without a body.
+Thus, in order to eagerly know what definitions *might* be executed at compile time and so need to be available including their bodies (in some executable shape), any definition used as an entry point to compile-time execution has to be tagged with the new `meta` modifier[^meta3].
+This is automatically done in built-in metaprogramming syntax such as `syntax`, `macro`, and `elab` but may need to be done explicitly when manually applying metaprogramming attributes such as `@[app_delab]`.
+
+[^meta3]: Semantically unrelated to the modifier of the same name in Lean 3.
+
+A `meta` definition may access (and thus invoke) any `meta` or non-`meta` definition of the current module.
+For accessing imported definitions, the definition must either have been marked as `meta` when it was declared or the import must be marked as such (`meta import` when the accessing definition is in the private scope and `public meta import` otherwise).
```

Comment:
A question I have when reading this is whether `meta import` is a superset of `import` or whether it's orthogonal.

Reply:
I'm not completely sure yet to be honest; it is currently implemented as a superset. But a `public meta import` would not necessarily have imply `public import` if the imported declarations are used only in the *bodies* of public declarations.

Reply:
I would think that my imports for purposes of metaprogramming/code generation should be reasonably distinct from my imports for purposes of generated run-time code, so that I could depend on a heavyweight library for proof automation but distribute an executable that's independent of it.